// graphql-js-tree/lib/Models/Spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Options = exports.Helpers = exports.Instances = exports.OperationType = exports.TypeExtension = exports.ValueDefinition = exports.TypeDefinition = exports.TypeSystemExtension = exports.TypeSystemDefinition = exports.Type = exports.Value = exports.Directive = exports.ScalarTypes = void 0;
var ScalarTypes;
(function (ScalarTypes) {
    ScalarTypes["Boolean"] = "Boolean";
    ScalarTypes["Float"] = "Float";
    ScalarTypes["ID"] = "ID";
    ScalarTypes["Int"] = "Int";
    ScalarTypes["String"] = "String";
})(ScalarTypes || (exports.ScalarTypes = ScalarTypes = {}));
var Directive;
(function (Directive) {
    Directive["SCHEMA"] = "SCHEMA";
    Directive["SCALAR"] = "SCALAR";
    Directive["OBJECT"] = "OBJECT";
    Directive["FIELD_DEFINITION"] = "FIELD_DEFINITION";
    Directive["ARGUMENT_DEFINITION"] = "ARGUMENT_DEFINITION";
    Directive["INTERFACE"] = "INTERFACE";
    Directive["UNION"] = "UNION";
    Directive["ENUM"] = "ENUM";
    Directive["ENUM_VALUE"] = "ENUM_VALUE";
    Directive["INPUT_OBJECT"] = "INPUT_OBJECT";
    Directive["INPUT_FIELD_DEFINITION"] = "INPUT_FIELD_DEFINITION";
})(Directive || (exports.Directive = Directive = {}));
var Value;
(function (Value) {
    Value["Variable"] = "Variable";
    Value["IntValue"] = "IntValue";
    Value["FloatValue"] = "FloatValue";
    Value["StringValue"] = "StringValue";
    Value["BooleanValue"] = "BooleanValue";
    Value["NullValue"] = "NullValue";
    Value["EnumValue"] = "EnumValue";
    Value["ListValue"] = "ListValue";
    Value["ObjectValue"] = "ObjectValue";
    Value["IDValue"] = "IDValue";
    Value["ScalarValue"] = "ScalarValue";
})(Value || (exports.Value = Value = {}));
var Type;
(function (Type) {
    Type["NamedType"] = "NamedType";
    Type["ListType"] = "ListType";
    Type["NonNullType"] = "NonNullType";
})(Type || (exports.Type = Type = {}));
var TypeSystemDefinition;
(function (TypeSystemDefinition) {
    TypeSystemDefinition["SchemaDefinition"] = "SchemaDefinition";
    TypeSystemDefinition["TypeDefinition"] = "TypeDefinition";
    TypeSystemDefinition["DirectiveDefinition"] = "DirectiveDefinition";
    TypeSystemDefinition["FieldDefinition"] = "FieldDefinition";
    TypeSystemDefinition["UnionMemberDefinition"] = "UnionMemberDefinition";
})(TypeSystemDefinition || (exports.TypeSystemDefinition = TypeSystemDefinition = {}));
var TypeSystemExtension;
(function (TypeSystemExtension) {
    TypeSystemExtension["SchemaExtension"] = "SchemaExtension";
    TypeSystemExtension["TypeExtension"] = "TypeExtension";
})(TypeSystemExtension || (exports.TypeSystemExtension = TypeSystemExtension = {}));
var TypeDefinition;
(function (TypeDefinition) {
    TypeDefinition["ScalarTypeDefinition"] = "ScalarTypeDefinition";
    TypeDefinition["ObjectTypeDefinition"] = "ObjectTypeDefinition";
    TypeDefinition["InterfaceTypeDefinition"] = "InterfaceTypeDefinition";
    TypeDefinition["UnionTypeDefinition"] = "UnionTypeDefinition";
    TypeDefinition["EnumTypeDefinition"] = "EnumTypeDefinition";
    TypeDefinition["InputObjectTypeDefinition"] = "InputObjectTypeDefinition";
})(TypeDefinition || (exports.TypeDefinition = TypeDefinition = {}));
var ValueDefinition;
(function (ValueDefinition) {
    ValueDefinition["EnumValueDefinition"] = "EnumValueDefinition";
    ValueDefinition["InputValueDefinition"] = "InputValueDefinition";
})(ValueDefinition || (exports.ValueDefinition = ValueDefinition = {}));
var TypeExtension;
(function (TypeExtension) {
    TypeExtension["ScalarTypeExtension"] = "ScalarTypeExtension";
    TypeExtension["ObjectTypeExtension"] = "ObjectTypeExtension";
    TypeExtension["InterfaceTypeExtension"] = "InterfaceTypeExtension";
    TypeExtension["UnionTypeExtension"] = "UnionTypeExtension";
    TypeExtension["EnumTypeExtension"] = "EnumTypeExtension";
    TypeExtension["InputObjectTypeExtension"] = "InputObjectTypeExtension";
})(TypeExtension || (exports.TypeExtension = TypeExtension = {}));
var OperationType;
(function (OperationType) {
    OperationType["query"] = "query";
    OperationType["mutation"] = "mutation";
    OperationType["subscription"] = "subscription";
})(OperationType || (exports.OperationType = OperationType = {}));
var Instances;
(function (Instances) {
    Instances["Argument"] = "Argument";
    Instances["Directive"] = "Directive";
    Instances["Implement"] = "Implement";
})(Instances || (exports.Instances = Instances = {}));
var Helpers;
(function (Helpers) {
    Helpers["Directives"] = "Directives";
    Helpers["Implements"] = "Implements";
    Helpers["Extend"] = "Extend";
    Helpers["Comment"] = "Comment";
})(Helpers || (exports.Helpers = Helpers = {}));
var Options;
(function (Options) {
    Options["array"] = "array";
    Options["required"] = "required";
    Options["name"] = "name";
})(Options || (exports.Options = Options = {}));
